File: src/ingestion/parser.py
import tree_sitter as tsjava
from tree_sitter import Language, Parser

JAVA_LANGUAGE = Language(tsjava.language())

class CodeParser:
    def __init__(self):
        self.parser = Parser(JAVA_LANGUAGE)

    def get_methods(self, file_path: str):
        """Extracts method blocks from a .java file using Tree-sitter."""
        with open(file_path, 'r', encoding='utf-8') as f:
            code_content = f.read()

        tree = self.parser.parse(bytes(code_content, 'utf8'))

        query = JAVA_LANGUAGE.query("""
            (method_declaration
                name: (identifier) @method_name
                body: (block) @method_body)
        """)

        captures = query.captures(tree.root_node)

        methods = []
        for node, tag in captures:
            if tag == "method_full":
                methods.append({
                    "text": node.text.decode('utf-8'),
                    "metadata": {
                        "file_path": file_path,
                        "start_line": node.start_point[0] + 1,
                        "end_line": node.end_point[0] + 1
                    }
                })
        return methods

File: src/ingestion/parser.ts
import * as Parser from 'web-tree-sitter';
import * as path from 'path';

export class JavaParser {
    private parser!: Parser;

    async initialize() {
        await Parser.init();
        this.parser = new Parser();
        // You'll need to download the tree-sitter-java.wasm file
        const Lang = await Parser.Language.load(path.join(__dirname, '../../parsers/tree-sitter-java.wasm'));
        this.parser.setLanguage(Lang);
    }

    /**
     * Extracts methods and classes as logical units (CAST strategy)[cite: 31, 32].
     */
    parseCode(code: string, filePath: string) {
        const tree = this.parser.parse(code);
        const chunks: any[] = [];

        // Query to find method declarations [cite: 14]
        const query = this.parser.getLanguage().query(`
            (method_declaration
                name: (identifier) @name
                body: (block) @body) @method
        `);

        const captures = query.captures(tree.rootNode);

        for (const capture of captures) {
            if (capture.name === 'method') {
                const node = capture.node;
                chunks.push({
                    content: code.substring(node.startIndex, node.endIndex),
                    metadata: {
                        source: filePath,
                        startLine: node.startPosition.row + 1,
                        endLine: node.endPosition.row + 1,
                        type: 'method'
                    }
                });
            }
        }
        return chunks;
    }
}